faiss/gpu/GpuIndexIVFFlat.cu

```diff
@@ -110,30 +172,44 @@ void GpuIndexIVFFlat::copyFrom(const faiss::IndexIVFFlat* index) {
 
     // The other index might not be trained
     if (!index->is_trained) {
-        FAISS_ASSERT(!this->is_trained);
+        FAISS_ASSERT(!is_trained);
         return;
     }
 
     // Otherwise, we can populate ourselves from the other index
-    FAISS_ASSERT(this->is_trained);
+    FAISS_ASSERT(is_trained);
 
     // Copy our lists as well
-    index_.reset(new IVFFlat(
+    set_index_(
             resources_.get(),
-            this->d,
-            this->nlist,
+            d,
+            nlist,
             index->metric_type,
             index->metric_arg,
             false,   // no residual
             nullptr, // no scalar quantizer
             ivfFlatConfig_.interleavedLayout,
             ivfFlatConfig_.indicesOptions,
-            config_.memorySpace));
-    baseIndex_ = std::static_pointer_cast<IVFBase, IVFFlat>(index_);
-    updateQuantizer();
-
-    // Copy all of the IVF data
-    index_->copyInvertedListsFrom(index->invlists);
+            config_.memorySpace);
+
+    if (config_.use_raft) {
+        printf("Reconstructing %d original vectors and adding to GPU index\n",
+               ntotal);
+
+        // Quantizer should already have been updated above. Add reconstructed
+        // vectors to raft index
+        if (ntotal > 0) {
+            std::vector<float> buf_host(ntotal * d);
+            std::vector<idx_t> ids(ntotal);
+            std::iota(ids.begin(), ids.end(), 0);
+            index->reconstruct_n(0, ntotal, buf_host.data());
```

Comment:
For non-flat index types (e.g., CPU IVF indices here), arbitrary user IDs can be assigned to vectors. There is no guarantee that they are sequential starting from 0.

I think we may need a "batch reconstruct all vectors and provide all indices stored in the index" function as well in order to support this case. Or, this may simply be a bad idea as the vectors may have already been quantized on the Faiss side of things using SQ or PQ, and then yet again we'd re-quantize on the RAFT side of things post-reconstruction. 

@mdouze  would be interested in your thoughts here too.

Reply:
Agree with Jeff here. Why not copy the data directly from the inverted lists? Like in 
https://github.com/facebookresearch/faiss/blob/868e17f29493075742170885f1f57c7b9e61d9ea/faiss/gpu/impl/IVFBase.cu#L325